### PR TITLE
Totp part4

### PR DESCRIPTION
**Pull request by 0x0ptr**

### CODE REVIEW

File: source_code/main_mcu/src/LOGIC/logic_gui.c [hunk 1]
```diff
@@ -303,24 +333,65 @@ void logic_gui_display_login_password(child_cred_node_t* child_node)
...
+    }
+
     /* Lines display settings */
-    cust_char_t* strings_to_be_displayed[2] = {child_node->login, child_node->cust_char_password};
-    uint16_t strings_y_positions[4] = {12, 36};
-    
+    uint16_t strings_y_positions[LOGIC_GUI_NUM_DISPLAY_CFG][LOGIC_GUI_NUM_LINES_MAX] = { {12, 36, 0}, {5, 20, 35} };
+
+    uint8_t num_lines_to_display = 1; //Login always displayed
+    num_lines_to_display += !!(child_node->passwordBlankFlag == FALSE); //Are we displaying password?
```
Comment (limpkin) [inline source_code/main_mcu/src/LOGIC/logic_gui.c:367]: Hmmm i'm not a big fan of this one...
could you maybe it macro like if you prefer? 
+= (child_node->passwordBlankFlag == FALSE)?0:1;

File: source_code/main_mcu/src/LOGIC/logic_gui.c [hunk 2]
```diff
@@ -303,24 +333,65 @@ void logic_gui_display_login_password(child_cred_node_t* child_node)
...
+    num_lines_to_display += !!(child_node->TOTP.TOTPsecretLen > 0);     //Are we displaying TOTP?
```
Comment (limpkin) [inline source_code/main_mcu/src/LOGIC/logic_gui.c:368]: same :)

File: source_code/main_mcu/src/LOGIC/logic_gui.c [hunk 3]
```diff
@@ -303,24 +333,65 @@ void logic_gui_display_login_password(child_cred_node_t* child_node)
...
+        strings_to_be_displayed[1] = TOTP_str;
+    }
+
+    /*
+     * Always set line 3. Either we only display line 1&2 and this is never used
+     * or we display all 3 lines and this will be used
+    */
+    strings_to_be_displayed[2] = TOTP_str;
+
+    uint8_t strings_y_pos_idx = !!(num_lines_to_display == LOGIC_GUI_NUM_LINES_MAX);
```
Comment (limpkin) [inline source_code/main_mcu/src/LOGIC/logic_gui.c:393]: same :)

File: source_code/main_mcu/src/LOGIC/logic_gui.c [hunk 4]
```diff
@@ -394,7 +475,7 @@ void logic_gui_display_login_password(child_cred_node_t* child_node)
                 if (scrolling_needed[i] != FALSE)
                 {
                     /* Scrolling required: display with the correct X offset */
-                    displayed_length = sh1122_put_string_xy(&plat_oled_descriptor, text_anim_x_offset[i], strings_y_positions[i], OLED_ALIGN_LEFT, strings_to_be_displayed[i], TRUE);
+                    displayed_length = sh1122_put_string_xy(&plat_oled_descriptor, text_anim_x_offset[i], strings_y_positions[strings_y_pos_idx][i], OLED_ALIGN_LEFT, strings_to_be_displayed[i], TRUE);
```
Comment (limpkin) [inline source_code/main_mcu/src/LOGIC/logic_gui.c:478]: very pretty.

**Review by limpkin — CHANGES_REQUESTED** [review]
very pretty PR (as usual)!
a few tiny changes :)

File: source_code/main_mcu/src/LOGIC/logic_gui.h [hunk 5]
```diff
@@ -29,13 +29,17 @@
...
 void logic_gui_clone_card(void);
 void logic_gui_change_pin(void);
 void logic_gui_erase_user(void);
 
-
-
-#endif /* LOGIC_GUI_H_ */
\ No newline at end of file
+/* Defines */
+#define LOGIC_GUI_NUM_LINES_MAX (3)       // Max number of lines to display for credentials
```
[the remark this answers is not in the document]
Comment (0x0ptr) [reply source_code/main_mcu/src/LOGIC/logic_gui.h:40]: :+1:

File: source_code/main_mcu/src/LOGIC/logic_gui.h [hunk 6]
```diff
@@ -29,13 +29,17 @@
...
+#define LOGIC_GUI_NUM_DISPLAY_CFG (2)     // Number of display cfg (2 or 3 lines)
```
[the remark this answers is not in the document]
Comment (0x0ptr) [reply source_code/main_mcu/src/LOGIC/logic_gui.h:41]: :+1:

File: source_code/main_mcu/src/LOGIC/logic_gui.c (hunk 1 above)
Comment (0x0ptr) [reply source_code/main_mcu/src/LOGIC/logic_gui.c:367]: you mean "+= (child_node->passwordBlankFlag == FALSE)? 1: 0;" :-)

That one always trips me up. Have you consider renaming it to passwordPresentFlag or something?

File: source_code/main_mcu/src/LOGIC/logic_gui.c (hunk 2 above)
Comment (0x0ptr) [reply source_code/main_mcu/src/LOGIC/logic_gui.c:368]: :+1:

File: source_code/main_mcu/src/LOGIC/logic_gui.c (hunk 3 above)
Comment (0x0ptr) [reply source_code/main_mcu/src/LOGIC/logic_gui.c:393]: :+1:

File: source_code/main_mcu/src/LOGIC/logic_gui.c (hunk 1 above)
Comment (limpkin) [reply source_code/main_mcu/src/LOGIC/logic_gui.c:367]: ohhhh right. this choice is because when the memory is erased, it is full of 0xFFFF...

**Pull request closed without merging**